scripts/validate_string_concatenation.py

```diff
@@ -1,51 +1,217 @@
+
+    For example, when this:
+
+    >>> foo = (
+    ...     "bar "
+    ...     "baz"
+    ... )
+
+    Is becoming this:
+
+    >>> foo = ("bar " "baz")
+
+    'Black' is not considering this as an
+    issue (see https://github.com/psf/black/issues/1051),
+    so we are checking it here instead.
+
+    Parameters
+    ----------
+    source_path : str
+        File path pointing to a single file.
+
+    Yields
+    ------
+    source_path : str
+        Source file path.
+    line_number : int
+        Line number of unconcatenated string.
+    msg : str
+        Explenation of the error.
+
+    Notes
+    -----
+    GH #30454
+    """
+    with open(source_path, "r") as file_name:
+        tokens: List = list(tokenize.generate_tokens(file_name.readline))
+
+    for current_token, next_token in zip(tokens, tokens[1:]):
+        if current_token[0] == next_token[0] == token.STRING:
+            yield (
+                source_path,
+                current_token[2][0],
+                (
+                    "String unnecessarily split in two by black. "
+                    "Please merge them manually."
+                ),
+            )
+
+
+def strings_with_wrong_placed_space(
+    source_path: str,
+) -> Generator[Tuple[str, int, str], None, None]:
+    """
+    Test case for leading spaces in concated strings.
+
+    For example:
+
+    >>> rule = (
+    ...    "We want the space at the end of the line, "
+    ...    "not at the beginning"
+    ... )
+
+    Instead of:
+
+    >>> rule = (
+    ...    "We want the space at the end of the line,"
+    ...    " not at the beginning"
+    ... )
+
+    Parameters
+    ----------
+    source_path : str
+        File path pointing to a single file.
+
+    Yields
+    ------
+    source_path : str
+        Source file path.
+    line_number : int
+        Line number of unconcatenated string.
+    msg : str
+        Explenation of the error.
+    """
+    with open(source_path, "r") as file_name:
+        tokens: List = list(tokenize.generate_tokens(file_name.readline))
+
+    for first_token, second_token, third_token in zip(tokens, tokens[1:], tokens[2:]):
+        # Checking if we are in a block of concated string
+        if (
+            first_token[0] == third_token[0] == token.STRING
+            and second_token[0] == token.NL
+        ):
+            # Striping the quotes
+            first_string = first_token[1][1:-1]
+            second_string = third_token[1][1:-1]
+
+            is_first_ends_single: bool = first_string.endswith(" ")
+            is_first_ends_newline: bool = first_string.endswith("\n")
+            is_second_starts_single: bool = second_string.startswith(" ")
+            is_second_starts_double: bool = second_string.startswith("  ")
+
+            if (
+                (not is_first_ends_single)
+                and is_second_starts_single
+                and (not is_second_starts_double)
+            ) and (not is_first_ends_newline):
```

Comment:
I don't think all this part is very easy to read. Probably worth creating functions for the conditions. Something like:
```
def single_space_in_wrong_place(line1, line2):
    return line1[-1] != " " and line2[0] == " " and line2[1] != " "
```

Reply:
What should the function return?  a tuple of (bool, str)

bool - indicating if the two strings passed/failed
str - the correct error message (ref https://github.com/pandas-dev/pandas/pull/30755#discussion_r365288596)

Reply:
I think just the bool will make the code better. See the example.